source/Ship.cpp

```diff
@@ -2907,14 +3147,42 @@ void Ship::Recharge(bool atSpaceport)
 
 bool Ship::CanRefuel(const Ship &other) const
 {
-	return (fuel - JumpFuel(targetSystem) >= other.JumpFuelMissing());
+	// Can't refuel if current ship has zero fuel capacity or other is full of fuel.
+	if(attributes.Get("fuel capacity") < 1 || GetSystem() != other.GetSystem() || IsEnemyInEscortSystem() || other.Fuel() == 1.)
+		return false;
+	if(CanBeCarried())
+	{
+		// Ensure all escorts have minimum one jump of fuel before refueling all
+		// escorts to maximum.  Only perform "max refueling" from within your
+		// own fleet.
+		if(IsYours() && !other.JumpFuelMissing())
+		{
+			// The escort fleet must have 1 jump minimum before refilling to 100% fuel.
+			if(!escortsHaveOneJump)
+				return false;
+			// Prioritize refueling non-carrier escorts before mission NPCs
+			if(!refuelMissionNpcEscort && !other.IsYours() && other.GetPersonality().IsEscort())
+				return false;
+		}
+		else if(!IsYours() && other.IsYours())
```

Comment:
So, just to be clear, because I saw the same kind of behaviour at another place, your accompanying escorts do not refuel you, but you refuel them if needed?